dependency/nomad_secure_variables_structs.go

```diff
@@ -0,0 +1,142 @@
+package dependency
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/hashicorp/nomad/api"
+)
+
+// NewNomadSVMeta is used to create a NomadSVMeta from a Nomad API
+// SecureVariableMetadata response.
+func NewNomadSVMeta(in *api.SecureVariableMetadata) *NomadSVMeta {
+	return &NomadSVMeta{
+		Namespace:   in.Namespace,
+		Path:        in.Path,
+		CreateIndex: in.CreateIndex,
+		ModifyIndex: in.ModifyIndex,
+		CreateTime:  nanoTime(in.CreateTime),
+		ModifyTime:  nanoTime(in.ModifyTime),
+	}
+}
+
+// NewNomadSecureVariable is used to create a NomadSecureVariable from a Nomad
+// API SecureVariable response.
+func NewNomadSecureVariable(in *api.SecureVariable) *NomadSecureVariable {
+	out := NomadSecureVariable{
+		Namespace:   in.Namespace,
+		Path:        in.Path,
+		CreateIndex: in.CreateIndex,
+		ModifyIndex: in.ModifyIndex,
+		CreateTime:  nanoTime(in.CreateTime),
+		ModifyTime:  nanoTime(in.ModifyTime),
+		Items:       map[string]NomadSVItem{},
+		sv:          in,
+	}
+
+	items := make(NomadSVItems, len(in.Items))
+	for k, v := range in.Items {
+		items[k] = NomadSVItem{k, v, &out}
+	}
+	out.Items = items
+	return &out
+}
+
+// NomadSecureVariable is a template friendly container struct that allows for
+// the NomadVar funcs to start inside of Items and have a rational way back up
+// to the SecureVariable that is JSON structurally equivalent to the API response.
+// This struct's zero value is not trivially usable and should be created with
+// NewNomadSecureVariable--especially when outside of the dependency package as
+// there is no access to sv.
+type NomadSecureVariable struct {
+	Namespace, Path          string
+	CreateIndex, ModifyIndex uint64
+	CreateTime, ModifyTime   nanoTime
+	Items                    NomadSVItems
+	sv                       *api.SecureVariable
+}
+
+func (cv NomadSecureVariable) Metadata() *NomadSVMeta {
+	return NewNomadSVMeta(cv.sv.Metadata())
+}
+
+type NomadSVItems map[string]NomadSVItem
+
+// Keys returns a sorted list of the Item map's keys.
+func (v NomadSVItems) Keys() []string {
+	out := make([]string, 0, len(v))
+	for k := range v {
+		out = append(out, k)
+	}
+	sort.Strings(out)
+	return out
+}
+
+// Values produces a key-sorted list of the Items map's values
+func (v NomadSVItems) Values() []string {
+	out := make([]string, 0, len(v))
+	for _, k := range v.Keys() {
+		out = append(out, v[k].String())
+	}
```

Comment:
This ends up allocating two slices instead of one because of the call to Keys. Can we do `for k, val := range v` instead? (Same applies to `Tuples()` below.)

Reply:
My main goal in using `Keys()` to create the other two outputs is that I wanted key-sorted order into the respective slices without repeating the sort code, so that they would range in the same order that go-template would order the map output.  Go template will do an implicit sort of keys so that they range consistently.

Reply:
Oh of course! That makes sense.